mlir/lib/Ion/Transforms/quantum_to_ion.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include "mlir/Dialect/Index/IR/IndexDialect.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Ion/IR/IonOps.h"
+#include "Ion/Transforms/Passes.h"
+#include "Ion/Transforms/Patterns.h"
+#include "Ion/Transforms/oqd_database_managers.hpp"
+#include "Ion/Transforms/oqd_database_types.hpp"
+#include "Quantum/IR/QuantumOps.h"
+
+using namespace mlir;
+using namespace catalyst::ion;
+
+namespace catalyst {
+namespace ion {
+
+#define GEN_PASS_DECL_QUANTUMTOIONPASS
+#define GEN_PASS_DEF_QUANTUMTOIONPASS
+#include "Ion/Transforms/Passes.h.inc"
+
+struct QuantumToIonPass : impl::QuantumToIonPassBase<QuantumToIonPass> {
+    using QuantumToIonPassBase::QuantumToIonPassBase;
+
+    LevelAttr getLevelAttr(MLIRContext *ctx, IRRewriter &builder, Level level){
+        return LevelAttr::get(ctx,
+            builder.getI64IntegerAttr(level.principal),
+            builder.getF64FloatAttr(level.spin),
+            builder.getF64FloatAttr(level.orbital),
+            builder.getF64FloatAttr(level.nuclear),
+            builder.getF64FloatAttr(level.spin_orbital),
+            builder.getF64FloatAttr(level.spin_orbital_nuclear),
+            builder.getF64FloatAttr(level.spin_orbital_nuclear_magnetization),
+            builder.getF64FloatAttr(level.energy)
+            );
+    }
+
+    TransitionAttr getTransitionAttr(MLIRContext *ctx, IRRewriter &builder, Transition transition){
+        return TransitionAttr::get(ctx,
+            getLevelAttr(ctx, builder, transition.level_0),
+            getLevelAttr(ctx, builder, transition.level_1),
+            builder.getF64FloatAttr(transition.einstein_a)
+            );
+    }
+
+    void runOnOperation() final
```

Comment:
Change to function pass (potentially would be reverted when we do frontend since each qnode function is in its own module now, but let's just put the big ion op in the function just to be safe)